packages/core/crates/core-packet/src/por.rs

```diff
@@ -0,0 +1,282 @@
+use core_crypto::derivation::{derive_ack_key_share, derive_own_key_share};
+use core_crypto::parameters::SECRET_KEY_LENGTH;
+use core_crypto::random::random_bytes;
+use core_crypto::types::{Challenge, CurvePoint, HalfKey, HalfKeyChallenge, PublicKey, Response};
+use utils_types::errors::GeneralError::ParseError;
+use utils_types::primitives::EthereumChallenge;
+use utils_types::traits::BinarySerializable;
+
+use crate::errors::{PacketError, Result};
+
+/// Proof of Relay secret length is twice the size of secp256k1 public key
+pub const POR_SECRET_LENGTH: usize = 2 * PublicKey::SIZE_COMPRESSED;
+
+/// Type that contains the challenge for the first ticket sent to the first relayer.
+#[derive(Clone)]
+#[cfg_attr(feature = "wasm", wasm_bindgen::prelude::wasm_bindgen(getter_with_clone))]
+pub struct ProofOfRelayValues {
+    pub ack_challenge: HalfKeyChallenge,
+    pub ticket_challenge: Challenge,
+    pub own_key: HalfKey,
+}
+
+impl ProofOfRelayValues {
+    /// Takes the secrets which the first and the second relayer are able to derive from the packet header
+    /// and computes the challenge for the first ticket.
+    pub fn new(secret_b: &[u8], secret_c: Option<&[u8]>) -> Self {
+        let s0 = derive_own_key_share(secret_b);
+        let s1 = derive_ack_key_share(secret_c.unwrap_or(&random_bytes::<SECRET_KEY_LENGTH>()));
+
+        Self {
+            ack_challenge: derive_ack_key_share(secret_b).to_challenge(),
+            ticket_challenge: Response::from_half_keys(&s0, &s1)
+                .expect("failed to derive response")
+                .to_challenge(),
+            own_key: s0,
+        }
+    }
+}
+
+/// Contains the Proof of Relay challenge for the next downstream node as well as the hint that is used to
+/// verify the challenge that is given to the relayer.
+#[derive(Clone)]
+pub struct ProofOfRelayString {
+    pub next_ticket_challenge: Challenge,
+    pub hint: HalfKeyChallenge,
+}
+
+impl ProofOfRelayString {
+    /// Creates instance from the shared secrets with node+2 and node+3
+    pub fn new(secret_c: &[u8], secret_d: Option<&[u8]>) -> Self {
+        assert_eq!(SECRET_KEY_LENGTH, secret_c.len(), "invalid secret length");
+        assert!(secret_d.is_none() || secret_d.unwrap().len() == SECRET_KEY_LENGTH);
+
+        let s0 = derive_ack_key_share(secret_c);
+        let s1 = derive_own_key_share(secret_c);
+        let s2 = derive_ack_key_share(secret_d.unwrap_or(&random_bytes::<SECRET_KEY_LENGTH>()));
+
+        Self {
+            next_ticket_challenge: Response::from_half_keys(&s1, &s2)
+                .expect("failed to derive response")
```

Comment:
Should this single failed instance kill the entire daemon?

Reply:
You're right, will remove it.